[jvm] Add jvm_options to platform; use from junit and run (#9150)
pantsbuild/pants: src/docs/common_tasks/multiple_jvm_versions.md

## Problem

You want to compile and/or run different test or binary targets using different JVMs.

Or, you need to specify command-line options globally for the JVM on a per-JVM version basis. For

> example, using `--add-opens` flags, but only when running on JVMs that support it.

Reviewer: I don't see `--add-opens` option added in this PR, did I miss anything? Possibly some example would be helpful.

Author: Yes. That's a good idea. `--add-opens` is a jvm option that was added in jdk 9, so it's incompatible with 8.

## Problem

You want to compile and/or run different test or binary targets using different JVMs.

Or, you need to specify command-line options globally for the JVM on a per-JVM version basis. For
example, using `--add-opens` flags, but only when running on JVMs that support it.

## Solution

> Declare JVM platforms using the `jvm-platform` configuration section to define the set of available

Reviewer 2: ```suggestion
In your global configuration (most likely `pants.toml`), declare JVM platforms using the `jvm-platform` configuration section to define the set of available
```

I like the continuity of "in X1, do Y1. In X2, do Y2". In this case, X1 = global configuration and X2 = your target.

platforms.

> On targets, add the kwarg `runtime_platform` set to the name of one of the platforms defined in

Reviewer 2: ```suggestion
On targets, add the named argument `runtime_platform` set to the name of one of the platforms defined in
```

`kwarg` is a Pythonic word, and a lot of users don't use Python that much.

Author: Thanks. I think that would be a good thing to have in a doc guideline doc somewhere.

> `jvm-platform`. That will tell pants to use that platform when running that target in a JVM.

Reviewer 2: It may be good to clarify here if you need to add it for _every_ target, or if not, what the default is.

To compile with a specific platform, set the `platform` on the target.

## Discussion

> 

Reviewer 2: An introductory sentence here would be good, to signal that "we're going to talk about what you can specify in the global config now". Something like:

```suggestion
Here are the different configurations that you can set when defining a JVM platform in your global configuration.

```

For compile time configuration, there are three main attributes, `source`, `target` and `args`.
`source` and `target` map directly to the javac arguments of the same name, though they accept

>  more aliases than javac. `args`, allows the platform to specify additional, global to that

Reviewer 2: > [...] more aliases than javac.

Examples here would be good, or even better: a pointer to a full list of aliases.

 platform, compile arguments.

For runtime configuration, `target` and `jvm_options` are used. `target` is used to determine the
version of the JVM to use. `jvm_options` is used by some tasks to allow platforms to specify
platform specific JVM options. These can make transitions between JVM versions smoother by

> allowing compatibility options to be provided globally.

Reviewer 2: I wonder if it's worth it to make this runtime/compile-time divide explicit in the structure of the options json. And how complicated that would be.

Author: oh, something like (ignoring quoting)

```
java8: {
  min: 1.8.0u40,
  max: 1.8.0u230,
  compile: {source: 8, target: 8, args: [-Dblah, -Xfoo]},
  runtime: {jvm_options:[-Xfoo -Xbar]}
}
```

Author: I think it's doable, but a bigger change.

An example pants.toml config might look like this:

    [jvm-platform]
    default_platform: java8
    default_runtime_platform: java10
    platforms =
    """
     {
        'java8': { 'source': '8', 'target': '8', 'args': [] },
        'java10': {'source': '10', 'target': '10', 'args': [],
                   'jvm_options': [
                       # --add-opens is a 9+ flag
                       '--add-opens=java.base/java.lang.reflect=ALL-UNNAMED'

>                    ] },

Reviewer: How does it interact with the existing `test.junit.jvm_options`?

For example,
`./pants test --test-junit-jvm-options["-Dabc=123"] <target>`. Is the final jvm_option merged with `--add-opens` assuming the target is 10?

Author: The options are folded into the runtime ones, before the target options. https://github.com/pantsbuild/pants/pull/9150/files#diff-8603ca893539bade8d782040a56d85b8L411-R415

Author: So in that case, you'd end up with `--add-opens ... -Dabc=123`

Reviewer: I see. Thanks for clarifying!

Reviewer 2: Mentioning that explicitly in the docs would be good :)
